### ISSUE TITLE
REP socket corrupts itself when REQ socket crashes

### ISSUE THREAD

**Issue by Snaipe** [nanomsg/nanomsg#756]
I have been chasing this bug for a while now, and finally have been able to isolate it out of my project.

In a context where multiple REQ clients connect to one REP server and send data back and forth, if data is flowing from one of the clients to the server while another client dies unexpectedly, the REP socket corrupts itself and starts sending nonsense back to the other REQ clients.

In a situation with 2 clients and 1 server, this would look like this:

```
Server spawns, REP socket bound to tcp://127.0.0.1:4224
Client 1 spawns, REQ socket connects to tcp://127.0.0.1:4224
Client 2 spawns, REQ socket connects to tcp://127.0.0.1:4224
Server calls recv
Client 1 calls send
Client 2 calls send
Server receives the message from Client 1, responds
Client 2 calls recv
Client 1 calls recv
Client 1 receives the response from Server
Server calls recv
Client 1 crashes
Server receives the message from Client 2
Server calls send
Client 2 receives garbage, and aborts.
```

This seems to manifest itself on Windows 10 with MSVC 2015 Update 2, built with both MSVC and MinGW-w64 (it's reproducible on Appveyor, notably). Otherwise, the issue may only happen sporadically (for instance, I cannot reproduce it on my Windows 10 VM with MSVC 2015 Update 2 for some reason).

Here is a [self-contained example that triggers the bug in question](https://ptpb.pw/jdx2).

The example itself could be minimized -- I only happened to get it to work to prove that the issue exists on nanomsg's side, and not on mine.

More tests might be needed to determine whether the socket can be recovered afterwards, or if the sent data might contain fragments of the original payload, but as I can only work on it over RDP on Appveyor because I otherwise cannot reproduce it locally, this is excruciatingly painful to do.

**Comment by Snaipe** [comment]
To add more info since the situation described above is approximative, here's the [log of a reproduction of the bug](https://ptpb.pw/-jbF) with 5 clients (4 process + 1 for the child sink) and 10 messages per client. I would have liked to be able to reproduce it with even less verbosity, but this seems to be the best that I can provide at the moment.

The above log was obtained with a [slightly modified version of the example ligned in the first post](https://ptpb.pw/W7tE)

**Comment by gdamore** [comment]
I have not been able to reproduce inside my Windows 8.1 VM.  Which makes debug hard.  Doing anything inside Windows is a substantial PITA.  Any chance at repro using POSIX code?

**Comment by gdamore** [comment]
Looking at the code, I don't see a likely problem.  What is interesting is your windows code is calling "spawn" -- I'm not familiar with the Windows process model -- do these processes share anything in common with their parent (ala fork)?

The xrep code should discard responses sent to a pipe that is no longer connected.  The way it does that is to look for the pipe id in the hash.  I'm pretty sure pipe IDs don't collide, and the allocation avoids reuse.  So I'm fairly stumped.

I'm wondering if there is some kind of data sharing taking place that we're not spotting (e.g. stack overrun, or something like that.)  Again, reproducible test cases would be very very helpful.

**Comment by gdamore** [comment]
One thing is that the hash table in used isn't locked.  In theory we should be operating in the context of the socket so it should be ok.  But I would admittedly feel lots better with explicit locks around data structures like the hash table.

An experiment you could add would be to add a mutex to the hash table and see if that solves the problem.

**Comment by Snaipe** [comment]
> I have not been able to reproduce inside my Windows 8.1 VM. Which makes debug hard. Doing anything inside Windows is a substantial PITA. Any chance at repro using POSIX code?

I've only been able to see it on Windows 10 -- on POSIX (and admitedly earlier versions of Windows) this works flawlessly. The problem might not come from REQ/REP per-se, but I haven't been able to reproduce with other protocols either.

> Looking at the code, I don't see a likely problem. What is interesting is your windows code is calling "spawn" -- I'm not familiar with the Windows process model -- do these processes share anything in common with their parent (ala fork)?

In the sample above, the created children have the same semantics as children that were spawned with fork+exec on POSIX -- although they don't share anything more than standard streams.

> An experiment you could add would be to add a mutex to the hash table and see if that solves the problem.

Will try that on my second round of tests.

The issue itself is hard to pinpoint. I suspect it might be caused by an implementation detail of the Windows-specific usock code that somehow impacted the REQ/REP code. Again, I'm not sure if this also impacts other protocols as I've only been able to spot it for REQ/REP. I'm also not sure if this impacts all transports, but so far it's reproducible with TCP and IPC.

**Comment by Snaipe** [comment]
After a second round of tests, I can confirm that locking the hash table did not help; I'm still having those assertions.

I'm not sure how the locking would have helped though. The children run in their own address space, so one of them crashing shouldn't have left the hash table of the server in an inconsistent state -- unless the fact that `handle_child_terminated` is threaded *is* the problem?

**Comment by gdamore** [comment]
I wonder if this is the issue with nn_realloc() not doing what we thought it did?

**Comment by Snaipe** [comment]
I'm not using nn_realloc in the self-contained example, is nn_chunk_realloc or similar used internally?

**Comment by JackDunaway** [comment]
I might have another type of repro scenario related to this.

It seems to be triggered when the REQ takes longer to reply than the REP timeout (e.g., if the REQ recv's a message, `sleep(200)`, then reply, where the REP timeout is 100msec). A crash happens when trying to close the REQ socket with the following trace:

```
 	ftw32.dll!nn_chunk_getptr(void * p) Line 211	C
 	ftw32.dll!nn_chunk_free(void * p) Line 150	C
 	ftw32.dll!nn_chunkref_term(nn_chunkref * self) Line 73	C
 	ftw32.dll!nn_req_term(nn_req * self) Line 110	C
>	ftw32.dll!nn_req_destroy(nn_sockbase * self) Line 130	C
 	ftw32.dll!nn_sock_shutdown(nn_fsm * self, int src, int type, void * srcptr) Line 916	C
 	ftw32.dll!nn_fsm_event_process(nn_fsm_event * self) Line 66	C
 	ftw32.dll!nn_ctx_leave(nn_ctx * self) Line 63	C
 	ftw32.dll!nn_worker_routine(void * arg) Line 204	C
 	ftw32.dll!nn_thread_main_routine(void * arg) Line 30	C
```

@Snaipe -- can you reproduce this anecdote, and do you think it's related to your issue here? I'm in early investigation right now, and will report back findings (perhaps in a new issue if it's a separate issue).

**Comment by JackDunaway** [comment]
PR #778 might be a minimal repro scenario or it may be a new bug -- what do you think, @Snaipe?

**Comment by Snaipe** [comment]
I'm not sure. In my example, there is no crash per-se, it's just that nn_recv on the REQ side fails with `EFSM` when the REP socket crashes. However, the fact that, in your example, a crash happens on the REQ side because the cancellation queue misbehaves might correlate the two issues. Maybe. Again, I'm really not sure, and I can't exactly test it right now, but if it's reproducible on appveyor only it's a good start.

**Comment by Snaipe** [comment]
After a bit of digging with samples, I'm managing to have `nn_recv` fail with EINTR on those specific cases, which is a bit weird since I did not trigger this condition before, but I'm left wondering if this is a programming error on my end or a windows-specific edge case.

In the meantime I'm going to try retrying the operation on EINTR.

**Comment by Snaipe** [comment]
Okay, retrying on EINTR loops indefinitely. It seems that once one of the client crashes the others are stuck in this EINTR condition.

**Comment by Snaipe** [comment]
Further research shows that https://github.com/nanomsg/nanomsg/blob/ad74d1a42cc24368999e938b01d327748f700871/src/utils/efd.c#L158 is triggered when the condition happens (with rc == ENOTSOCK. Incidentally, rc can never be EINTR according to nn_err_wsa_to_posix), which makes in turn nn_sock_recv return -EINTR.

Finally, some progress!

**Comment by Snaipe** [comment]
Okay, I found what caused this.

It turns out that this isn't a nanomsg bug, but a race condition in how the library was used within multiple threads. `RegisterWaitForSingleObject` would spawn a new thread for each dead child to collect its status. In the self-contained example, `send_payload` would be called, which would in turn call `nn_send` then `nn_recv` on a REQ socket.

The issue becomes apparent when two threads try to do these operations without an additional layer of synchronization. The "bug" I reported was only the manifestation of the following sequence of operations:

1. [thread#1] nn_send on REQ
2. [thread#2] nn_send on REQ, cancels first nn_send
3. [thread#1] nn_recv on REQ, gets ack from 2nd nn_send
4. [thread#2] nn_recv on REQ, no pending request so return with errno = EFSM

I can't believe it took me that long to realize that requests did not queue and that it was a stupid race condition.
